commands/board/attach.go

```diff
@@ -0,0 +1,160 @@
+ * Copyright 2018 ARDUINO SA (http://www.arduino.cc/)
+ *
+ * This software is released under the GNU General Public License version 3,
+ * which covers the main part of arduino-cli.
+ * The terms of this license can be found at:
+ * https://www.gnu.org/licenses/gpl-3.0.en.html
+ *
+ * You can be released from the requirements of the above licenses by purchasing
+ * a commercial license. Buying such a license is mandatory if you want to modify or
+ * otherwise use the software for commercial activities involving the Arduino
+ * software without disclosing the source code of your own applications. To purchase
+ * a commercial license, send an email to license@arduino.cc.
+ */
+
+package board
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/arduino/arduino-cli/arduino/cores"
+	"github.com/arduino/arduino-cli/arduino/cores/packagemanager"
+	"github.com/arduino/arduino-cli/arduino/sketches"
+	"github.com/arduino/arduino-cli/commands"
+	"github.com/arduino/arduino-cli/rpc"
+	discovery "github.com/arduino/board-discovery"
+	paths "github.com/arduino/go-paths-helper"
+)
+
+func BoardAttach(ctx context.Context, req *rpc.BoardAttachReq, taskCB commands.TaskProgressCB) (*rpc.BoardAttachResp, error) {
+
+	pm := commands.GetPackageManager(req)
+	if pm == nil {
+		return nil, errors.New("invalid instance")
+	}
+	var sketchPath *paths.Path
+	if req.GetSketchPath() != "" {
+		sketchPath = paths.New(req.GetSketchPath())
+	}
+	sketch, err := sketches.NewSketchFromPath(sketchPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening sketch: %s", err)
+	}
+
+	boardURI := req.GetBoardURI()
+	fqbn, err := cores.ParseFQBN(boardURI)
+	if err != nil && !strings.HasPrefix(boardURI, "serial") {
+		boardURI = "serial://" + boardURI
+	}
+
+	if fqbn != nil {
+		sketch.Metadata.CPU = sketches.BoardMetadata{
+			Fqbn: fqbn.String(),
+		}
+	} else {
+		deviceURI, err := url.Parse(boardURI)
+		if err != nil {
+			return nil, fmt.Errorf("invalid Device URL format: %s", err)
+		}
+
+		var findBoardFunc func(*packagemanager.PackageManager, *discovery.Monitor, *url.URL) *cores.Board
+		switch deviceURI.Scheme {
+		case "serial", "tty":
+			findBoardFunc = findSerialConnectedBoard
+		case "http", "https", "tcp", "udp":
+			findBoardFunc = findNetworkConnectedBoard
+		default:
+			return nil, fmt.Errorf("invalid device port type provided")
+		}
+
+		duration, err := time.ParseDuration(req.GetSearchTimeout())
+		if err != nil {
+			duration = time.Second * 5
+		}
+
+		monitor := discovery.New(time.Second)
+		monitor.Start()
+
+		time.Sleep(duration)
+
+		// TODO: Handle the case when no board is found.
+		board := findBoardFunc(pm, monitor, deviceURI)
+		if board == nil {
+			return nil, fmt.Errorf("no supported board found at %s", deviceURI.String())
+		}
+		taskCB(&rpc.TaskProgress{Name: "Board found: " + board.Name(), Completed: true})
+
+		sketch.Metadata.CPU = sketches.BoardMetadata{
+			Fqbn: board.FQBN(),
+			Name: board.Name(),
+		}
```

Comment:
We should probably stop the monitor: when running as a pure CLI it doesn't matter because when the CLI exits the monitor is released by the OS, but when running as a daemon I don't know if the resource is leaked.

Please put a `TODO` this is related to board discovery and `board list` command.

Reply:
done